### DISCUSSION TITLE
Configure JWT Authentication with env variables

### DISCUSSION THREAD

**Discussion by MacherelR** [emqx/emqx#12093] [Q&A] [answered]
Hello, I'd like to set up and configure jwt authentication in an emqx deployment using later on helm and the deployment file. I specifically do not want to hard-write the secret used for the tokens inside the deployment file as I have them saved in an encrypted location... Therefore I'd rather set an env variable that'll contain the secret and configure to use that env variable.
Here's my actual file:
~~~
# Emqx install
{{- if .Values.emqx.enabled }}
apiVersion: apps.emqx.io/v2beta1
kind: EMQX
metadata:
  name: emqx
spec:
  image: emqx:5.1
  imagePullPolicy: Always
  config:
    data: |
      listeners.tcp.test {
         bind = "0.0.0.0:1884"
         max_connections = 1024000
      }
      authentication = [{
         from = password
         mechanism = jwt
         enable = true
         secret = <SECRET_ENV_VAR>
         secret_base64_encoded = false
         algorithm = hmac-based
         use_jwks = false
         verify_claims {
            username = "${username}"
         }
         acl_claim_name = "acl"
      }]
  listenersServiceTemplate:
    spec:
      type: LoadBalancer
  dashboardServiceTemplate:
    spec:
      type: LoadBalancer
  env:
    tokenSecret: {{.Values.emqx.env.tokenSecret}}
{{- end }}
~~~
Basically, how do I tell emqx to use the tokenSecret env var as secret in the configuration ?

**Comment by lafirest** [comment]
@MacherelR 
I think this `EMQX_AUTHENTICATION__1__SECRET ` will work.
See the [Environment Variables](https://www.emqx.io/docs/en/latest/configuration/configuration.html#environment-variables) here for more.

**Reply by MacherelR** [reply]
It works thank you ! :)

**Reply by mgabeler-lee-6rs** [reply]
Is there a trick here? I'm trying to use this and emqx startup fails with:
```
2023-12-18T20:34:48.175672+00:00 [error] failed_to_check_schema: emqx_conf_schema
2023-12-18T20:34:48.177925+00:00 [error] #{kind => validation_error,path => "authentication.1",reason => "missing_mechanism_field"}
ERROR: call_hocon_failed: -v -t 2023.12.18.20.34.47 -s emqx_conf_schema -c /opt/emqx/etc/emqx.conf -d /opt/emqx/data/configs generate
```
Which seems strange because I definitely have the mechanism field in the config file.

**Reply by thalesmg** [reply]
Hi @mgabeler-lee-6rs .

Which EMQX version are you using?

Also, do you have more than one authn entry in your config? In particular, more than one _not separated with a comma_?  i.e.:

```hcl
authentication = [
  { ... }  # no comma here
  { ... }
]
```

If you are using 5.1 and you do have more than one config without a separating comma, please try to add a comma like:


```hcl
authentication = [
  { ... },  # N.B. the comma
  { ... }
]
```

**Reply by mgabeler-lee-6rs** [reply]
I'm using the `emqx/emqx` docker container, so latest release -- so v5.3.2. I have just one authn entry in my config.

From some experimentation, it seems that setting the environment variable causes it to try to completely overwrite the entire authn object, instead of just the one property within it.

Given that, this bit in the release notes from v5.0.12 sounds vaguely suspicious, but it's such an old release I'm skeptical if it's related: https://github.com/emqx/emqx/blob/master/changes/v5.0.12-en.md?plain=1#L39-L43

Lastly, in case it's relevant, the authn entry is defined in my `cluster.hocon` file, not in the `emqx.conf` file.

<details>
<summary>Here is the whole `cluster.hocon` file</summary>

```
authentication = [
	{
		mechanism = jwt
		enable = true
		from = password
		algorithm = hmac-based
		# secret = "..." set from environment
		secret_base64_encoded = false
		use_jwks = false
		verify_claims {
			aud = emqx
			sub = "${username}"
		}
	}
]
authorization {
	cache {
		enable = false
	}
	deny_action = disconnect
	no_match = deny
}
bridges {
	webhook {
		# todo
	}
}
rule_engine {
	ignore_sys_message = true
	jq_function_default_timeout = 10s
	rules {
		# todo
	}
}

api_key {
	bootstrap_file = "/opt/emqx/data/configs/api-keys"
}

cluster {
	autoclean = 5m
	# TODO: why can't we put cluster discovery settings here?
	#discovery_strategy = k8s
	#k8s {
	#	service_name = "emqx"
	#	namespace = "default"
	#	apiserver = "https://kubernetes.default.svc:443"
	#	service_name = "emqx"
	#}
}
```
</details>

You can see some comments about some other "fun" I'm having in there with the cluster discovery settings, but that's a Different Issue (I think?)

**Reply by thalesmg** [reply]
Thanks for the detailed information.

The error is a bit confusing indeed.  The reason for it seemingly "dropping" the other fields when overriding a single one is that `cluster.hocon` is not considered when generating the initial configuration (`call_hocon ... generate` in the error message).  Thus, at that step, when setting only that field via environment variables, indeed only that one is available. 

`cluster.hocon` is generated when you change configurations using the API and synchronized between the cluster nodes.  So, if you created the authn config via the HTTP API, you'll need to change the secret using the same methods or editing the file directly before starting the node, I'm afraid.

Support for specifying file-based secrets was very recently added for bridges (will be available in 5.4.0), but it's not available yet for that particular authentication mechanism, unfortunately.  It might be added in future versions.

**Reply by thalesmg** [reply]
Alternatively, if you really want to override the secret using env vars, then you could _move_ (i.e., don't leave any authn config in `cluster.hocon`) the authn config from `cluster.hocon` to `etc/emqx.conf` and then use the env vars to set/override certain fields.

Note: if you go down that route, then you shouldn't set/update such configurations via the HTTP API/dashboard.  You'll only be able to change them by editing the file/env var config in your deploy and restarting the whole node.

**Reply by mgabeler-lee-6rs** [reply]
Thanks for the explanation.  I think the docs could use some improvement here in this case however, as they regularly describe configuration as layering emqx.conf < cluster.hocon < environment.

For context. I'm using cluster.hocon via a k8s configmap to bring in my own config without completely replacing the base emqx.conf from the docker image, but it sounds like that's ... not a good idea here. It also may explain why I can't get the cluster discovery to work from within cluster.hoconn:) I'll work out a sensible replacement, and if I have more questions, they clearly merit a separate thread

**Reply by Frtrillo** [reply]
> Alternatively, if you really want to override the secret using env vars, then you could _move_ (i.e., don't leave any authn config in `cluster.hocon`) the authn config from `cluster.hocon` to `etc/emqx.conf` and then use the env vars to set/override certain fields.
> 
> Note: if you go down that route, then you shouldn't set/update such configurations via the HTTP API/dashboard. You'll only be able to change them by editing the file/env var config in your deploy and restarting the whole node.

so what is the best practice to only load username and password from ENV variables instead of hard coding them in the file...? Currently I load them from cluster.hocon as well before building the docker image but would rather prefer to load username and password from env variables but then i have this issue as well as it doesnt load the rest of authn config from cluster.hocon

**Reply by zmstone** [reply, marked as answer]
## The reported issue

I believe this was due to the fact that `cluster.hocon` was not loaded when generating the base config in 5.3.
With only one variable defined, the node would not boot, but fail with error like below:

```
2024-10-19T13:13:47.130009+00:00 [error] failed_to_check_schema: emqx_conf_schema
2024-10-19T13:13:47.134331+00:00 [error] #{kind => validation_error,path => "authentication.1",reason => "missing_mechanism_field"}
ERROR: call_hocon_failed: -v -t 2024.10.19.13.13.45 -s emqx_conf_schema -c /opt/emqx/etc/emqx.conf -d /opt/emqx/data/configs generate
```

This has been fixed in since 5.6.0, see my test on 5.8.1 for example:
<img src="https://github.com/user-attachments/assets/c0debcb4-abbe-422b-bb4d-e2ff96bb20c1" style="box-shadow: 0 0 10px rgba(0, 0, 0, 0.3);" width="600">

## About document

config overriding rules are documented here:
https://docs.emqx.com/en/emqx/latest/configuration/configuration.html#configure-override-rules

## Password best-practice

~The best practice to avoid storing secrets in cluster.hocon or emqx.conf is to use `file://</path/to/secret` for the config value string.~
~this is supported since 5.4.0~

Sorry, `file://` secret is not yet supported for JWT authentication.
